packages/contracts/src/contracts/current/forwarder/MixinForwarderCore.sol

```diff
@@ -0,0 +1,145 @@
+            uint256 feeRecipientFeeAmount = safeDiv(safeMul(remainingTakerTokenAmount, feeProportion), PERCENTAGE_DENOMINATOR);
+            remainingTakerTokenAmount = safeSub(remainingTakerTokenAmount, feeRecipientFeeAmount);
+            feeRecipient.transfer(feeRecipientFeeAmount);
+        }
+        return remainingTakerTokenAmount;
+    }
+
+    /// @dev Withdraws the remaining WETH, deduct and pay fees from this amount based on the takerTokenAmount to the feeRecipient.
+    ///      If a user overpaid ETH initially, the fees are calculated from the amount traded and deducted from withdrawAmount.
+    ///      Any remaining ETH is sent back to the user.
+    /// @param withdrawAmount The amount to withdraw from the WETH contract.
+    /// @param takerTokenAmount The total amount that was transacted in WETH, fees are calculated from this value.
+    /// @param feeProportion The proportion of fees
+    /// @param feeRecipient The recipient of the fees
+    /// @return remainingTakerTokenAmount the amount remaining after fees have been deducted and paid to feeRecipient.
+    function withdrawPayAndDeductFee(
+        uint256 withdrawAmount,
+        uint256 takerTokenAmount,
+        uint16 feeProportion,
+        address feeRecipient)
+        internal
+        returns (uint256 remainingTakerTokenAmount)
+    {
+        // Return all of the excess WETH if any after deducting fees on the amount
+        if (withdrawAmount > 0) {
+            ETHER_TOKEN.withdraw(withdrawAmount);
+            remainingTakerTokenAmount = withdrawAmount;
+            // Fees proportional to the amount traded
+            uint256 feeAmount = safeSub(takerTokenAmount, payAndDeductFee(takerTokenAmount, feeProportion, feeRecipient));
+            remainingTakerTokenAmount = safeSub(remainingTakerTokenAmount, feeAmount);
+            if (remainingTakerTokenAmount > 0) {
+                msg.sender.transfer(remainingTakerTokenAmount);
+            }
+        }
+        return remainingTakerTokenAmount;
+    }
+
+    /// @dev Buys feeAmount of ZRX fee tokens, taking into account the fees on buying fee tokens. This will guarantee
+    ///      At least feeAmount of ZRX fee tokens are purchased (sometimes slightly over due to rounding issues).
+    ///      It is possible that a request to buy 200 ZRX fee tokens will require purchasing 202 ZRX tokens
+    ///      As 2 ZRX is required to purchase the 200 ZRX fee tokens. This guarantees at least 200 ZRX for future purchases.
+    /// @param orders An array of Order struct containing order specifications for fees.
+    /// @param signatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeAmount The number of requested ZRX fee tokens.
+    /// @return totalFillResults Amounts filled and fees paid by maker and taker. makerTokenAmount is the zrx amount deducted of fees
+    function buyFeeTokensInternal(
```

Comment:
Why does buying fee tokens differ from buying an exact number of ERC20 tokens? This looks a lot like the market buy in the exchange core, which is called by `buyExactTokensInternal`.

Reply:
The exchange rate is different when buying ZRX and there are ZRX fees. If we don't take this into account than fees are removed and a 100 BUY of ZRX returns 98 ZRX (after fees). If this happens theres not enough to fill other orders (ERC721 for example) are we are short